### PR TITLE
Add Recent Timestep Scheduling Improvements to DDIM Inverse Scheduler

### PR DESCRIPTION
**Pull request by clarencechen**
# What does this PR do?
- Adds Zero SNR scaling and trailing timestep specification functionality to `DDIMInverseScheduler`
- Roll timesteps by one to reflect origin-destination semantic discrepancy relative to `DDIMScheduler`
- Restore `set_alpha_to_one` option to handle negative origin timesteps at first inversion step
- Remove `set_alpha_to_zero` option not used due to previous truncation of timestep array in pipelines
- Update pipelines to invert last denoising step (as the first inversion step) as pointed out in PR #3436

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/diffusers/blob/main/CONTRIBUTING.md)?
- [x] Did you read our [philosophy doc](https://github.com/huggingface/diffusers/blob/main/PHILOSOPHY.md) (important for complex PRs)?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/diffusers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_diffedit.py [hunk 1]
```diff
@@ -1246,9 +1246,9 @@ def invert(
 
         # 7. Noising loop where we obtain the intermediate noised latent image for each timestep.
         num_warmup_steps = len(timesteps) - num_inference_steps * self.inverse_scheduler.order
-        inverted_latents = [latents.detach().clone()]
+        inverted_latents = []
         with self.progress_bar(total=num_inference_steps - 1) as progress_bar:
-            for i, t in enumerate(timesteps[:-1]):
+            for i, t in enumerate(timesteps):
```
Comment (patrickvonplaten) [inline src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_diffedit.py:1251]: Why do we need to do these changes?

File: src/diffusers/schedulers/scheduling_ddim_inverse.py [hunk 2]
```diff
@@ -124,18 +169,20 @@ def __init__(
         beta_schedule: str = "linear",
         trained_betas: Optional[Union[np.ndarray, List[float]]] = None,
         clip_sample: bool = True,
-        set_alpha_to_zero: bool = True,
+        set_alpha_to_one: bool = True,
```
Comment (patrickvonplaten) [inline src/diffusers/schedulers/scheduling_ddim_inverse.py:172]: This is backwards breaking, why do this change?

File: src/diffusers/schedulers/scheduling_ddim_inverse.py [hunk 3]
```diff
@@ -200,12 +251,29 @@ def set_timesteps(self, num_inference_steps: int, device: Union[str, torch.devic
             )
 
         self.num_inference_steps = num_inference_steps
-        step_ratio = self.config.num_train_timesteps // self.num_inference_steps
-        # creates integer timesteps by multiplying by ratio
-        # casting to int to avoid issues when num_inference_step is power of 3
-        timesteps = (np.arange(0, num_inference_steps) * step_ratio).round().copy().astype(np.int64)
+
+        # "leading" and "trailing" corresponds to annotation of Table 1. of https://arxiv.org/abs/2305.08891
+        if self.config.timestep_spacing == "leading":
```
Comment (patrickvonplaten) [inline src/diffusers/schedulers/scheduling_ddim_inverse.py:256]: ok for me!

**Comment by patrickvonplaten** [comment]
Thanks for the PR @clarencechen . However, a couple of things would break current setups. Can we try to not change config names? Also it seems like you are correcting a bug here? Can you add a reproducible code snippet that shows that there is a bug?

File: src/diffusers/schedulers/scheduling_ddim_inverse.py (hunk 2 above)
Comment (clarencechen) [reply src/diffusers/schedulers/scheduling_ddim_inverse.py:172]: `set_alpha_to_zero` should be captured in `kwargs`, and retrieved to display the deprecation message.

**Comment by patrickvonplaten** [comment]
Actually I think we should go forward with this PR. The `DDIMScheduler` also has the `set_alpha_to_one` argument, so it'd be good to align this here. 

@sayakpaul @linoytsaban what do you think about the PR?

**Comment by patrickvonplaten** [comment]
Also related: https://github.com/huggingface/diffusers/pull/3831

**Comment by sayakpaul** [comment]
> @sayakpaul @linoytsaban what do you think about the PR?

I concur with your thoughts:

> Actually I think we should go forward with this PR. The DDIMScheduler also has the set_alpha_to_one argument, so it'd be good to align this here.

**Comment by clarencechen** [comment]
@patrickvonplaten @sayakpaul Just added some new fast tests to ensure the inverse schedulers work across the available config settings - I'll verify the slow pipeline tests involving inversion tomorrow.

**Review by patrickvonplaten — APPROVED** [review]
Good for me! 

@sayakpaul wdyt?

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_diffedit.py [hunk 4]
```diff
@@ -1037,7 +1037,7 @@ def generate_mask(
         )
 
         # 4. Preprocess image
-        image = preprocess(image).repeat_interleave(num_maps_per_mask, dim=0)
+        image = self.image_processor.preprocess(image).repeat_interleave(num_maps_per_mask, dim=0)
```
Comment (sayakpaul) [inline src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_diffedit.py:1040]: Great reuse!

Should we then maybe just delete `preprocess()` to help reduce the LOC from the pipeline source?

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_diffedit.py [hunk 5]
```diff
@@ -1246,9 +1246,9 @@ def invert(
 
         # 7. Noising loop where we obtain the intermediate noised latent image for each timestep.
         num_warmup_steps = len(timesteps) - num_inference_steps * self.inverse_scheduler.order
-        inverted_latents = [latents.detach().clone()]
-        with self.progress_bar(total=num_inference_steps - 1) as progress_bar:
-            for i, t in enumerate(timesteps[:-1]):
+        inverted_latents = []
```
Comment (sayakpaul) [inline src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_diffedit.py:1249]: Where is this variable being used now?

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_diffedit.py [hunk 6]
```diff
@@ -1336,7 +1336,7 @@ def __call__(
         self,
         prompt: Optional[Union[str, List[str]]] = None,
         mask_image: Union[torch.FloatTensor, PIL.Image.Image] = None,
-        image_latents: torch.FloatTensor = None,
+        image_latents: Union[torch.FloatTensor, PIL.Image.Image] = None,
```
Comment (sayakpaul) [inline src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_diffedit.py:1339]: `image_latents` is not a good name when we're accepting PIL images too. `image` works for me.

**Review by sayakpaul — APPROVED** [review]
Amazing work!

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_diffedit.py (hunk 6 above)
Comment (patrickvonplaten) [reply src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_diffedit.py:1339]: Agree, happy to deprecate the name via `**kwargs`
